docs/architecture/grpc-for-wcf-developers/channel-credentials.md

```diff
@@ -0,0 +1,153 @@
+
+Kestrel can be configured to require a client certificate, and optionally to carry out some validation of the supplied certificate before accepting incoming connections. This configuration is done in the `CreateWebHostBuilder` method of the `Program` class, rather than in `Startup`.
+
+```csharp
+public static IHostBuilder CreateHostBuilder(string[] args) =>
+    Host.CreateDefaultBuilder(args)
+        .ConfigureWebHostDefaults(webBuilder =>
+        {
+            var serverCert = ObtainServerCertificate();
+            webBuilder.UseStartup<Startup>()
+                .ConfigureKestrel(kestrelServerOptions => {
+                    kestrelServerOptions.ConfigureHttpsDefaults(opt =>
+                    {
+                        opt.ClientCertificateMode = ClientCertificateMode.RequireCertificate;
+
+                        // Verify that client certificate was issued by same CA as server certificate
+                        opt.ClientCertificateValidation = (certificate, chain, errors) =>
+                            certificate.Issuer == serverCert.Issuer;
+                    });
+                });
+        });
+
+```
+
+The `ClientCertificateMode.RequireCertificate` setting will cause Kestrel to immediately reject any connection request that does not provide a client certificate, but it will not validate the certificate. Adding the `ClientCertificateValidation` callback enables Kestrel to validate the client certificate (in this case, ensuring that it was issued by the same Certificate Authority as the server certificate) at the point the connection is made, before the ASP.NET Core pipeline is engaged.
+
+### Adding ASP.NET Core Certificate authentication
+
+Certificate authentication is provided by the Microsoft.AspNetCore.Authentication.Certificate package ([www.nuget.org/packages/Microsoft.AspNetCore.Authentication.Certificate](https://www.nuget.org/packages/Microsoft.AspNetCore.Authentication.Certificate)).
+
+Add the Certificate authentication service in `ConfigureServices`, and add Authentication and Authorization to the ASP.NET Core pipeline in `Configure`.
+
+```csharp
+public class Startup
+{
+    private readonly bool _isDevelopment;
+
+    public Startup(IWebHostEnvironment env)
+    {
+        _isDevelopment = env.IsDevelopment();
+    }
+
+    public void ConfigureServices(IServiceCollection services)
+    {
+        services.AddAuthentication(CertificateAuthenticationDefaults.AuthenticationScheme)
+            .AddCertificate(options =>
+            {
+                if (_isDevelopment)
+                {
+                    // DO NOT DO THIS IN PRODUCTION!
+                    options.RevocationMode = X509RevocationMode.NoCheck;
+                }
+            });
+        services.AddAuthorization();
+        services.AddGrpc();
+    }
+
+    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
+    {
+        app.UseAuthentication();
+        app.UseAuthorization();
+
+        app.UseRouting();
```

Comment:
UseRouting should come before authn/authz middleware.

Reply:
Oops.